jgrapht-core/src/main/java/org/jgrapht/GraphMetrics.java

```diff
@@ -207,6 +209,114 @@
                 && (graph.getType().isDirected() || graph.getType().isAllowingMultipleEdges());
         return girth;
     }
+
+    /**
+     * An $O(|V|^3)$ naive implementation for counting non-trivial triangles in an undirected graph
+     * induced by the subset of vertices.
+     *
+     * @param graph the input graph
+     * @param vertexSubset the vertex subset
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return the number of triangles in the graph induced by vertexSubset
+     */
+    private static <V, E> long naiveCountTriangles(Graph<V, E> graph, List<V> vertexSubset){
+        long total = 0;
+
+        for (int i = 0; i < vertexSubset.size(); i++) {
+            for (int j = i + 1; j < vertexSubset.size(); j++) {
+                for (int k = j + 1; k < vertexSubset.size(); k++) {
+                    V u = vertexSubset.get(i);
+                    V v = vertexSubset.get(j);
+                    V w = vertexSubset.get(k);
+
+                    if (graph.containsEdge(u, v) && graph.containsEdge(v, w) && graph.containsEdge(w, u)){
+                        total++;
+                    }
+                }
+            }
+        }
+
+        return total;
+    }
+
+    /**
+     * An $O(|E|^{3/2})$ algorithm for counting the number of non-trivial triangles in an undirected graph.
+     * A non-trivial triangle is formed by tree distinct vertices all connected to each others.
+     *
+     * <p>
+     * For more details of this algorithm see Ullman, Jeffrey: "Mining of Massive Datasets", Cambridge University Press,
+     * Chapter 10
+     *
+     * @param graph the input graph
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return the number of triangles in the graph
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws IllegalArgumentException if {@code graph} is not undirected
+     */
+    public static <V, E> long getNumberOfTriangles(Graph<V, E> graph){
+        GraphTests.requireUndirected(graph);
+
+        final int sqrtV = (int)Math.sqrt(graph.vertexSet().size());
+
+        List<V> vertexList = new ArrayList<>(graph.vertexSet());
+
+        /*
+            The book suggest the following comparator: "Compare vertices based on their degree.
+            If equal compare them of their actual value, since they are all integers".
+
+            They are not integers in this implementation so we compare them on their hashCode.
+         */
+        Comparator<V> comparator = Comparator.comparingInt(graph::degreeOf).thenComparingInt(Object::hashCode);
```

Comment:
Not sure whether this is the right approach. The second comparison is "lexicographic" which you usually perform by memory addresses, at least in languages like C/C+. You should generate a unique number for each vertex at the beginning of the algorithm and use that instead. See `BoruvkaMinimumSpanningTree` for an example.

The problem with `hashCode` is that it is implementation specific (depending on the JRE) and you are not sure that your vertices will not collide. They probably won't but still, the correctness cannot rely on such assumptions.

Reply:
Or use System.identityHashCode, which is similar to a C++ memory address.

Reply:
Fixed. Thank you @jsichi for the suggestion.

Reply:
While better since the user cannot override, I do not think that there is any guarantee for uniqueness. 

If for some reason you get the same hashCode for two distinct vertices, you cannot distinguish between them and thus might count the triangles wrong. 

You need lexicographic ordering if you are going to argue that you count each triangle only once.

Reply:
This is a classic case where the JVM spec does not state any implementation details, except the fact that you must always return the same number when invoked with the same object. 

However, a valid implementation of hashCode could always return the number zero for every object. 

Regarding memory addresses, I think they no longer use that technique. But even if they did, since the JVM moves objects around, you are forced to cache the initial address (as you must always return the same hashCode) and thus two objects might get the same address and thus hashCode. 

The following link contains some useful info (https://srvaroa.github.io/jvm/java/openjdk/biased-locking/2017/01/30/hashCode.html)

Reply:
I have switched to something similar to `BoruvkaMinimumSpanningTree`. Thank you for the information, @d-michail!

Reply:
Sorry for the bad advice...I acquired that misconception somewhere along the way.

Reply:
It's ok...I'm glad I learnt something new about `hashCode`.